src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java

```diff
@@ -174,18 +171,31 @@ public InnerClassLambdaMetafactory(MethodHandles.Lookup caller,
                                implKind == MethodHandleInfo.REF_invokeSpecial ||
                                implKind == MethodHandleInfo.REF_invokeStatic && implClass.isHidden();
         int parameterCount = factoryType.parameterCount();
+        ClassDesc[] argDescs;
+        MethodTypeDesc constructorTypeDesc;
         if (parameterCount > 0) {
-            argNames = new String[parameterCount];
             argDescs = new ClassDesc[parameterCount];
             for (int i = 0; i < parameterCount; i++) {
-                argNames[i] = "arg$" + (i + 1);
                 argDescs[i] = classDesc(factoryType.parameterType(i));
             }
+            constructorTypeDesc = MethodTypeDescImpl.ofValidated(CD_void, argDescs);
         } else {
-            argNames = EMPTY_STRING_ARRAY;
             argDescs = EMPTY_CLASSDESC_ARRAY;
+            constructorTypeDesc = MTD_void;
+        }
+        this.argDescs = argDescs;
+        this.constructorTypeDesc = constructorTypeDesc;
+    }
+
+    private static String argName(int i) {
+        String argName = i < ARG_NAME_CACHE_SIZE ? ARG_NAME_CACHE[i] : null;
+        if (argName == null) {
+            argName = "arg$".concat(Integer.toString(i + 1));
```

Comment:
With `Integer.toString` in the mix perhaps just as well keep this as a standard concat?
```suggestion
            argName = "arg$" + (i + 1);
```

Reply:
New bytecode in last commit's `computeArgName`:
```
         0: ldc           #186                // String arg$
         2: iload_0
         3: iconst_1
         4: iadd
         5: invokestatic  #188                // Method java/lang/Integer.toString:(I)Ljava/lang/String;
         8: invokevirtual #193                // Method java/lang/String.concat:(Ljava/lang/String;)Ljava/lang/String;
        11: areturn

```
Old bytecode in constructor:
```
       222: iload         11
       224: new           #150                // class java/lang/StringBuilder
       227: dup
       228: invokespecial #152                // Method java/lang/StringBuilder."<init>":()V
       231: ldc           #155                // String arg$
       233: invokevirtual #157                // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
       236: iload         11
       238: iconst_1
       239: iadd
       240: invokevirtual #161                // Method java/lang/StringBuilder.append:(I)Ljava/lang/StringBuilder;
       243: invokevirtual #164                // Method java/lang/StringBuilder.toString:()Ljava/lang/String;
       246: aastore
       247: aload_0
```
I believe using concat is cleaner.

Reply:
Irrelevant; idiomatic concatenation is more concise in source and easier to read, comprehend and check, so should generally be preferred. That the legacy javac concat translatiom is lousy for constant + primitive is a different matter.